packages/protocol/contracts/L1/ProverPool.sol

```diff
@@ -0,0 +1,409 @@
+                }
+            }
+            emit Slashed(addr, amountToSlash);
+        }
+    }
+
+    function stake(
+        uint64 amount,
+        uint32 rewardPerGas,
+        uint32 maxCapacity
+    )
+        external
+        nonReentrant
+    {
+        // Withdraw first
+        _withdraw(msg.sender);
+        // Force this prover to fully exit
+        _exit(msg.sender, true);
+        // Then stake again
+        if (amount != 0) {
+            _stake(msg.sender, amount, rewardPerGas, maxCapacity);
+        } else if (rewardPerGas != 0 || maxCapacity != 0) {
+            revert INVALID_PARAMS();
+        }
+    }
+
+    function exit() external nonReentrant {
+        _withdraw(msg.sender);
+        _exit(msg.sender, true);
+    }
+
+    // Withdraws staked tokens back from matured an exit
+    function withdraw() external nonReentrant {
+        if (!_withdraw(msg.sender)) revert NO_MATURE_EXIT();
+    }
+
+    // Returns a staker's information
+    function getStaker(address addr)
+        public
+        view
+        returns (Staker memory staker, Prover memory prover)
+    {
+        staker = stakers[addr];
+        if (staker.proverId != 0) {
+            unchecked {
+                prover = provers[staker.proverId];
+            }
+        }
+    }
+
+    // Returns the pool's current total capacity
+    function getCapacity() public view returns (uint256 capacity) {
+        unchecked {
+            for (uint256 i = 1; i <= MAX_NUM_PROVERS; ++i) {
+                capacity += provers[i].currentCapacity;
+            }
+        }
+    }
+
+    function getProvers()
+        public
+        view
+        returns (Prover[] memory _provers, address[] memory _stakers)
+    {
+        _provers = new Prover[](MAX_NUM_PROVERS);
+        _stakers = new address[](MAX_NUM_PROVERS);
+        for (uint256 i; i < MAX_NUM_PROVERS; ++i) {
+            _provers[i] = provers[i + 1];
+            _stakers[i] = proverIdToAddress[i + 1];
+        }
+    }
+
+    function getProverWeights(uint32 feePerGas)
+        public
+        view
+        returns (
+            uint256[MAX_NUM_PROVERS] memory weights,
+            uint32[MAX_NUM_PROVERS] memory erpg
+        )
+    {
+        Prover memory _prover;
+        unchecked {
+            for (uint32 i; i < MAX_NUM_PROVERS; ++i) {
+                _prover = provers[i + 1];
+                if (_prover.currentCapacity != 0) {
+                    // Keep the effective rewardPerGas in [75-125%] of feePerGas
+                    if (_prover.rewardPerGas > feePerGas * 125 / 100) {
+                        erpg[i] = feePerGas * 125 / 100;
+                    } else if (_prover.rewardPerGas < feePerGas * 75 / 100) {
+                        erpg[i] = feePerGas * 75 / 100;
+                    } else {
+                        erpg[i] = _prover.rewardPerGas;
+                    }
+                    weights[i] = _calcWeight(_prover.stakedAmount, erpg[i]);
+                }
```

Comment:
A proposal for this item:
"Another desired feature is to somehow encourage provers to submit proofs ASAP. This feature remains to be discussed and potential implemented."  https://github.com/taikoxyz/taiko-mono/issues/14090 

The final reward per gas is also related to the proof time. The faster the proof is provided, the higher the reward. This means that the weight is not related to proof time, but once a prover is selected, we encourage it to submit their proof as soon as possible in order to receive a greater reward.

Reply:
We'll work on this one later :)